Allows sitemap output on root of public folder

pedrolamas opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.dev/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)
-->

<!--
  Is this a blog post? Check out the docs at https://www.gatsbyjs.com/contributing/blog-contributions/, and please mention if the blog post is pre-approved
  by someone from Gatsby.
-->

## Description

<!-- Write a brief description of the changes introduced by this PR -->
All this PR does is allow developers to set the `output` option on the `gatsby-sitemap-plugin` to `""` (empty string) so that the files can be written to the root of the public folder.

Some (like me) might prefer that instead of creating a subfolder to store the files, plus this mitigates a separate issue where the index file always points to the root folder instead of the specified output folder.

### Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Gatsby Internals docs.
  - If no docs exist:
    - Create a stub for documentation including bullet points for how to use the feature, code snippets (including from happy path tests), etc.
  - Tag @gatsbyjs/documentation for review, pairing, polishing of the documentation
-->
https://www.gatsbyjs.com/plugins/gatsby-plugin-sitemap/

## Related Issues

<!--
  Link to the issue that is fixed by this PR (if there is one)
  e.g. Fixes #1234

  Link to an issue that is partially addressed by this PR (if there are any)
  e.g. Addresses #1234

  Link to related issues (if there are any)
  e.g. Related to #1234
-->
Fixes #31095

pedrolamas commented:
FYI, I've tested this locally and works fine with an empty string but will fail with `null` value, hence why I didn't put that as allowed!

KyleAMathews commented:
Curious why an empty string vs `/`?

pedrolamas commented:
@KyleAMathews I tested with "/" and noticed that the header that was created had double slashes in it (so it showed "//sitemap-index.xml"), using a blank string fixed that!

KyleAMathews commented:
Ah ok, that's the bug that should be fixed then here — https://github.com/gatsbyjs/gatsby/blob/a2a9dc9e45edc6b82e9109418f4e63b3fd7a96e0/packages/gatsby-plugin-sitemap/src/gatsby-ssr.js#L20

When joining paths, it's best to use `path.join` as that then takes care automatically of any possible doubled slashes e.g:

```shell
➜  gatsby-plugin-sitemap git: ✗ node
Welcome to Node.js v16.0.0.
Type ".help" for more information.
> const path = require(`path`)
undefined
> path.join(`/`, `/`)
'/'
```

pedrolamas commented:
I do wonder if this issue is showing in other bits, given this: https://github.com/gatsbyjs/gatsby/blob/a2a9dc9e45edc6b82e9109418f4e63b3fd7a96e0/packages/gatsby-plugin-sitemap/src/gatsby-ssr.js#L5-L6

moonmeister commented:
I was going to summit the pr to convert to `path.join` but it probably is a better solution than an empty string. You're welcome to do it here.

pedrolamas commented:
AFAIK, `path.join()` should not be used for this, as we are talking about urls and not actual paths, and in Windows land, that can causes even bigger problems:

![image](https://user-images.githubusercontent.com/85504/116785662-b09da280-aa92-11eb-83d3-7a7cb03492e5.png)

Having said that, `path.posix.join()` works, but once again, that is NOT the correct way to do this!

pedrolamas commented:
This looks more like it!

https://github.com/gatsbyjs/gatsby/blob/54d4721462b9303fed723fdcb15ac5d72e103778/deprecated-packages/gatsby-plugin-guess-js/src/gatsby-ssr.js#L7-L12

But I do wonder if there's already some dependency that Gatsby is using to do this correctly?

moonmeister commented:
Good call, stark over flow looks like it has some good suggestions. https://stackoverflow.com/questions/16301503/can-i-use-requirepath-join-to-safely-concatenate-urls

pedrolamas commented:
Given the comments from @KyleAMathews, I've changed the code to fix when `output: "/"` to ensure that the link is correctly created.

I've also added a small test to ensure this keeps the expected behavior.

moonmeister commented:
Thanks for the fix and especially for the tests.

moonmeister reviewed (approved):
I haven't run the code but looks good.

KyleAMathews commented:
Ah yeah, `posix.join` is what we want cause windows. We use that elsewhere in core https://github.com/gatsbyjs/gatsby/search?q=posix.join

Nice work @pedrolamas !

KyleAMathews commented:
Published in `gatsby-plugin-sitemap@4.1.0-next.1` (so you can install it with `gatsby-plugin-sitemap@next`)
